File: RenderServer/src/main/java/me/hydos/alchemytools/network/netty/MessageDecoder.java
package me.hydos.alchemytools.network.netty;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ReplayingDecoder;
import me.hydos.alchemytools.network.RenderServer;

import java.io.IOException;
import java.util.List;

public class MessageDecoder extends ReplayingDecoder<ChannelMessage> {

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
        try {
            var packetId = in.readShort();
            var direction = in.readByte();
            var dataLength = in.readInt();
            var packetBytes = new byte[dataLength];
            for (int i = 0; i < packetBytes.length; i++) packetBytes[i] = in.readByte();
            out.add(new ChannelMessage(packetId, direction, packetBytes));
        } catch (Exception e) {
            RenderServer.LOGGER.error("Failed to decode packet :(");
            throw new RuntimeException(e);
        }
    }
}

File: RenderClient/netty/MessageDecoder.cs
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;

namespace RenderClient.netty;

public class MessageDecoder : ReplayingDecoder<ChannelMessage> {

    public MessageDecoder() : base(new ChannelMessage()) {
    }

    protected override void Decode(IChannelHandlerContext context, IByteBuffer input, List<object> output) {
        var packetId = input.ReadShort();
        var direction = input.ReadByte();
        var packetBytes = new byte[input.Capacity - input.ReaderIndex];
        input.GetBytes(input.ReaderIndex, packetBytes);
        output.Add(new ChannelMessage(packetId, direction, packetBytes));
    }
}